bolt/runtime/instr.cpp

```diff
@@ -566,12 +566,98 @@ FunctionDescription::FunctionDescription(const uint8_t *FuncDesc) {
 /// Read and mmap descriptions written by BOLT from the executable's notes
 /// section
 #if defined(HAVE_ELF_H) and !defined(__APPLE__)
+#define BUF_SIZE 1024
+#define NAME_MAX 256
+#define ADDR_SIZE 32
+
+void *__attribute__((noinline)) __get_pc() {
+  return __builtin_extract_return_addr(__builtin_return_address(0));
+}
+
+/// Get full path to the real binary by getting current virtual address
+/// and searching for the appropriate link in address range in
+/// /proc/self/map_files
+static char *getBinaryPath() {
+  const char dirPath[] = "/proc/self/map_files/";
+  static char target_path[NAME_MAX] = {};
+  char buf[BUF_SIZE], start[ADDR_SIZE], end[ADDR_SIZE];
+  char findBuf[NAME_MAX];
+  char *strcp;
+  int lenS, lenE;
+  long nread;
+  unsigned long curAddr, saddr, eaddr;
+  struct dirent *d;
+
+  if (target_path[0] != '\0')
+    return target_path;
+
+  curAddr = (unsigned long)__get_pc();
+  uint64_t FDdir = __open(dirPath,
+                          /*flags=*/0 /*O_RDONLY*/,
+                          /*mode=*/0666);
+  assert(static_cast<int64_t>(FDdir) > 0,
+         "failed to open /proc/self/map_files");
+
+  for (;;) {
```

Comment:
Here is an example at the end of the page https://man7.org/linux/man-pages/man2/getdents.2.html and it seems like we need this loop (getdents can read to the buf only BUF_SIZE elements at once), otherwise we might lose some entries.

Reply:
Sorry, I missed the line

```
     if (nread == 0)
      break;
```

with the exit condition. So it reads until nread is zero. Then perhaps this outer loop could be written as:

  ```
while (nread = __getdents(FDdir, (struct dirent *)buf, BUF_SIZE) ) {
  }
```

To make the exit condition more explicit